core/src/test/scala/io/isomarcte/http4s/active/requests/core/ActiveRequestMiddlewareTest.scala

```diff
@@ -0,0 +1,88 @@
+package io.isomarcte.http4s.active.requests.core.unit
+
+import cats.data._
+import cats.effect._
+import cats.implicits._
+import io.isomarcte.http4s.active.requests.core._
+import java.util.concurrent._
+import org.http4s._
+import org.http4s.server._
+import scala.concurrent._
+
+final class ActiveRequestMiddlewareTest extends BaseTest {
+
+  "serviceUnavailableMiddleware" should "reject requests if there are too many concurrently running" in io {
+    val ec: ExecutionContext = BaseTest.cachedEC
+    val limit: Int = 1
+    val latch: CountDownLatch = new CountDownLatch(2)
+    val semaphore: Semaphore = new Semaphore(limit)
+    val request: Request[IO] = Request[IO]()
+    val f: IO[Unit] = IO(latch.countDown()) *> IO(semaphore.acquire())
+    val (activeRequests, middleware): (IO[Long], HttpMiddleware[IO]) =
+      ActiveRequestMiddleware.serviceUnavailableMiddleware[IO, Long](
+        limit.toLong
+      )
+    val service: HttpService[IO] = middleware(ActiveRequestMiddlewareTest.effectService[IO](f))
```

Comment:
WARN: Is this the way you expect the library consumer to work with this, or is this construct just available for testing? I would suggest that the public interface should not have any ability to query the state, and should simply return the `HttpMiddleware` value. Do you see this differently?

Reply:
I do see it differently. Being able to query the state is very important for things like metrics.

Reply:
I'm open to suggestions on alternative ways to encode that too.